iota/api.py

```diff
@@ -432,6 +432,21 @@ def broadcast_and_store(self, trytes):
     """
     return extended.BroadcastAndStoreCommand(self.adapter)(trytes=trytes)
 
+  def check_consistency(self, tails):
+    # type: (Iterable[TransactionHash]) -> dict
+    """
+    Checks transaction hashes for promotability
+
+    :return:
+      Dict containing the following::
+          {
+            'state': bool,
```

Comment:
Based on the JS library and the IRI code (https://github.com/iotaledger/iri/blob/501e9493bc0cbd1ef1b3e0df6da39efb8ef4083d/src/main/java/com/iota/iri/service/dto/CheckConsistency.java) this value appears to just be a single `bool`. I would have expected a list as well though based on some of the other API functions but that doesn't seem to be the case...

Reply:
Huh, that's really curious.  Thanks for the link!

Looking at the source of [`checkConsistencyStatement()`](https://github.com/iotaledger/iri/blob/501e9493bc0cbd1ef1b3e0df6da39efb8ef4083d/src/main/java/com/iota/iri/service/API.java#L360-L395), it seems like the command indeed treats all of the transactions as an atomic unit.

Also, it looks like the response should have an `info` attribute in addition to `state`.

Reply:
Ok, I found some more info on how promoting works:

From <https://iota.stackexchange.com/a/801/283>:

> And how does the wallet choose which button to show? In latest iri (1.4.1.4), a function `checkConsistency` was added which you can call on a pending transaction (or more of them) and it will tell you if it is still possible for this transaction (or all the transactions simultaneously if you give more than one) to be confirmed, or not (because it conflicts with another already confirmed transaction). When the transaction is still consistent, you can only promote it (since it gives you a better chance of getting confirmed than reattachment and avoids creating land-mines), and if not, you can only reattach it (since promoting it would never confirm; there is already a land-mine so the new one will not become one; and it cannot happen that the reattachment is attached to the original transaction since tip selection will never choose inconsistent transactions).

Also, have a look at this:

https://www.reddit.com/r/Iota/comments/7kysio/the_new_promote_system/

Reply:
I'll add some of that information to the docstring. As for `info`, it looks like it should be in the response, but when I print out the resulting dictionary from `check_consistency` I see something like the following:

```python
{'state': True, 'duration': 239}
```

It looks like `info` starts as `null` and only gets set when `state` is `false`. If it's `null` when [`gson.toJson`](https://github.com/iotaledger/iri/blob/501e9493bc0cbd1ef1b3e0df6da39efb8ef4083d/src/main/java/com/iota/iri/service/API.java#L923) is called, that [value will be ignored and not included in the response dictionary](https://stackoverflow.com/questions/3923759/gson-ignoring-map-entries-with-value-null).

Are there any existing examples in the code base for how to document this behavior?

Reply:
Hmm.  Interesting.

I feel like the output from the IRI should be consistent.  I think I'll log an issue on the IRI issue tracker about this.

In the meantime, I think this is the first time we've run into this situation; what are your thoughts about how the Python code should handle it?

Reply:
The two options I can think of are:

1. Keep it as is and just document that the `info` field will only be there when `state` is `False`. If this changes on the IRI we would have to update the documentation, but it probably wouldn't cause any API changes in this library.

2. Have this library modify the resulting dictionary with something like `response['info'] = None` or `response['info'] = ''` when the `info` key isn't there. However, this could cause breaking API changes for this library if for example we set `info` to `None` when empty and later the IRI decides to initialize `info` as an empty string and return `''` when `state` is `False`.

To me, option 1 feels like the more correct thing to do even if it is a bit frustrating until the IRI is fixed.

Reply:
Make sense.  Okay, let's go with option 1 then!